src/encoded/static/components/report.js

```diff
@@ -282,53 +283,190 @@ ColumnHeader.propTypes = {
+    firstColumnTitle: PropTypes.string.isRequired, // Title of first column
+};
+
+
+// Displays a modal dialog with every possible column for the type of object being displayed.
+// This lets you choose which columns you want to appear in the report.
+class ColumnSelector extends React.Component {
+    constructor(props) {
+        super(props);
+
+        // Set the initial React states.
+        this.state = {
+            columns: props.columns, // Make (effectively) a local mutatable copy of the columns
+            sortOption: 'default', // Default sorting option
+        };
+
+        // Bind `this` to non-React methods.
+        this.submitHandler = this.submitHandler.bind(this);
         this.toggleColumn = this.toggleColumn.bind(this);
+        this.handleSelectAll = this.handleSelectAll.bind(this);
+        this.handleSelectOne = this.handleSelectOne.bind(this);
+        this.handleSortChange = this.handleSortChange.bind(this);
     }
 
-    toggle(e) {
-        e.preventDefault();
-        this.setState(prevState => ({
-            open: !prevState.open,
-        }));
+    toggleColumn(columnPath) {
+        // Called every time a column's checkbox gets clicked on or off in the modal. `columnPath`
+        // is the query-string term corresponding to each column. First, if the column is getting
+        // turned off, make sure we have at least one other column selected, because at least one
+        // column has to be selected.
+        if (this.state.columns[columnPath].visible) {
+            // The clicked column is currently visible, so before we make it invisible, make sure
+            // at least one other column is also visible.
+            const allColumnKeys = Object.keys(this.state.columns);
+            const anotherVisible = allColumnKeys.some(key => key !== columnPath && this.state.columns[key].visible);
+            if (!anotherVisible) {
+                // A checkbox is being turned off, and no other checkbox is checked, so ignore the
+                // click.
+                return;
+            }
+        }
+
+        // Either a checkbox is being turned on, or it's being turned off and another checkbox is
+        // still checked. Change the component state to reflect the new checkbox states. Presumably
+        // if the setState callback returned no properties setState becomes a null op, so the above
+        // test could be done inside the setState callback. The React docs don't say what happens
+        // if you return no properties (https://reactjs.org/docs/react-component.html#setstate) so
+        // I avoided doing this.
+        this.setState((prevState) => {
+            // Toggle the `visible` state corresponding to the column whose checkbox was toggled.
+            // Then set that as the new React state which causes a redraw of the modal with all the
+            // checkboxes in the correct state.
+            prevState.columns[columnPath].visible = !prevState.columns[columnPath].visible;
```

Comment:
https://reactjs.org/docs/react-component.html#setstate says:

> prevState is a reference to the previous state. It should not be directly mutated. Instead, changes should be represented by building a new object based on the input from prevState and props.

which means we should not be directly mutating prevState.columns

Reply:
Yes, I missed that. I’ll change all those.